### ISSUE TITLE
App health stuck at "Progressing" forever when pod has `restartPolicy: Never`

### ISSUE THREAD

**Issue by lindhe** [argoproj/argo-cd#15317]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When an Application deploys a pod that has `restartPolicy: Never`, the Application gets stuck with health "Progressing" seemingly forever.

**To Reproduce**

Apply this manifest:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  name: debug
  namespace: argocd
spec:
  destination:
    server: https://kubernetes.default.svc
    namespace: default
  source:
    path: manifests
    repoURL: https://github.com/lindhe/debug-argocd-15317.git
```

That app installs just this seemingly innocuous pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: foo
spec:
  containers:
    - image: busybox
      command: ['sh', '-c', 'sleep 3600']
      name: bar
  restartPolicy: Never
```

**Expected behavior**

The application health should soon become "Healthy".

**Screenshots**

<img width="1080" alt="Screenshot 2023-09-01 083537" src="https://github.com/argoproj/argo-cd/assets/7773090/24d80128-05a6-4b06-8498-ccefa687ca7a">

**Version**

```shell
argocd: v2.7.9+0ee33e5
  BuildDate: 2023-07-24T18:26:12Z
  GitCommit: 0ee33e52dd1f1bb944488584fc6f854b929f1180
  GitTreeState: clean
  GoVersion: go1.19.11
  Compiler: gc
  Platform: linux/amd64
WARN[0000] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web.
argocd-server: v2.8.0+804d4b8
```

**Logs**

No relevant logs that I've found.

**Other information**

Similar to #5620 and some other issues found by searching for "health progressing".

I'm running an RKE2 cluster with version `v1.26.6+rke2r1` and Cilium `1.13.2` installed.

**Comment by blakepettersson** [comment]
This seems to be by [design](https://github.com/argoproj/argo-cd/issues/7933#issuecomment-993949395), or rather there is nothing currently better than what there is (see the [relevant snippet](https://github.com/argoproj/gitops-engine/blob/187312fe86ed627b1cf4147c13dccf9ff331424b/pkg/health/health_pod.go#L30) in gitops engine).

**Comment by lindhe** [comment]
Ah, that's interesting. Thanks for finding it!

It sounds to me like ignoring all pods that sets `restartPolicy: Never` or `restartPolicy: OnFailure` is a very blunt tool to achieve those ends. Perhaps we can find a better way of doing things?

After reading the documentation, it is my understanding that all resource hooks have the `argocd.argoproj.io/hook` annotation. Perhaps we could therefore instead look for pods that fulfills these requirements:

* has `restartPolicy: Never` or `restartPolicy: OnFailure`.
* has the `argocd.argoproj.io/hook` annotation _or_ is owned by an object with that annotation.

Do you think it sounds reasonable to look for an alternative solution like this, in order to avoid these unexpected behaviors?

**Comment by blakepettersson** [comment]
> Perhaps we could therefore instead look for pods that fulfills these requirements:

From my perspective that does sound reasonable. I suspect that there might be a caveat or two to this which would still make this a no-go. @jessesuen do you have any input regarding this?

**Comment by blakepettersson** [comment]
@lindhe perhaps the way to move this forward is to come in one of the Argo CD Contributors meetings and make a proposal?

**Comment by lindhe** [comment]
Sure, I'd love to. Where do I find the schedule?

**Comment by blakepettersson** [comment]
You can see the agenda and Zoom link [here](https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit). TLDR every Thursday at 17.15 CET

**Comment by lindhe** [comment]
Great. I don't have time this week, but expect me soon! :)

**Comment by nweisenauer-sap** [comment]
Hey there, any updates on this one? @lindhe

**Comment by lindhe** [comment]
Sorry, I dropped the ball on this one. 😕 Q4 last year was hectic for me. Still plan to push for it on the contributors meeting, but if someone has time to join and discuss it before I have time, go for it! 👍

**Comment by jpuyy** [comment]
I also met this problem for flink task manager, it may affect the flink autoscale.

flink operator release-1.7.0

**Comment by billylaing** [comment]
Thank y'all for an awesome tool! I just ran into this issue today and found my way here. 

I have ArgoCD as an app in my ArgoCD app set and I am experiencing this issue with two pods with `restartPolicy: Never`:

`pod/argo-cd-redis-ha-configmap-test`
`pod/argo-cd-redis-ha-service-test`

It is minor but would be great to have it report healthy since everything does when I look at `kubectl -n argocd get all`.

Let me know if I could assist in some way. Thanks!

**Comment by jkleinlercher** [comment]
It seems kubevirt vm-launcher pods are also affected. https://github.com/kubevirt/kubevirt/blob/ea53cc9d444227a033c55d521979e6ccc688456f/pkg/virt-controller/services/template.go#L583

**Comment by ErwinJapie** [comment]
Any updates on this issue?

**Comment by pyang55** [comment]
Im waiting on an update on this as well

**Comment by blakepettersson** [comment]
No updates, you can safely assume that unless anyone has stated otherwise that this is still up for grabs and that there's no ETA for the resolution of this issue. Help is always welcome! 🙏

**Comment by jkleinlercher** [comment]
@blakepettersson I am just wondering why no one from Redhat tries to solve this because OpenShift Virtualization is affected by this problem and they push OpenShift Virtualization very much. See https://github.com/kubevirt/kubevirt/issues/11813.
 @terrytangyuan

**Comment by Liammarwood** [comment]
Are there any workarounds for this? Does anyone have an example custom heath check we can implement?

**Comment by lindhe** [comment]
> Are there any workarounds for this? Does anyone have an example custom heath check we can implement?

Yes. The work-around is to not set `restartPolicy: Never` unless its in a Pod spawned by a Job (which is supposed to be marked as "Progressing" until the pod has terminated).

**Comment by jkleinlercher** [comment]
Problem is that kubevirt uses it .. then it should get fixed in https://github.com/kubevirt/kubevirt/issues/11813

**Comment by lindhe** [comment]
Cool, I hadn't heard about a real-world application with this use-case before! That is interesting. My original example was literally just created as a dummy example to see if Argo CD worked, and I didn't imagine there being legitimate use-cases for `restartPolicy: Never` outside of jobs.

Can you, please, summarize how `kubevirt` makes use of it? I can read in the issue that it's the "virt-launcher", but I have no idea what that is. It's a one-time Pod but not a Job??

**Comment by jkleinlercher** [comment]
It is a permanent running pod. Why it has restartPolicy never .. I don’t know. I would hope that some Redhat folks would have a look at it since they are heavy contributors in ArgoCD AND kubevirt and use both products in their OpenShift Virtualization offering.

**Comment by Liammarwood** [comment]
I wouldn't say not setting Restart policy to Never is a work around,  a lot of operators are doing this outside of our powers. My example of seeing this is using Nifikop, but from other forums I know there are a few more of them.

**Comment by EuGras** [comment]
> I wouldn't say not setting Restart policy to Never is a work around, a lot of operators are doing this outside of our powers. My example of seeing this is using Nifikop, but from other forums I know there are a few more of them.

yes, same with Apache Flink deployments managed by Flink Operator

**Comment by Liammarwood** [comment]
Maybe we can implement an annotation to stop this behaviour? Then we wouldn't have to break anything current?

**Comment by surajEE** [comment]
Facing similar issue while installing runner scale set for Github's Actions Runner Controller as we don't have control over restartPolicy of gha-runner-scale-set-listener. Waiting for the solution/workaround.

**Comment by webertrlz** [comment]
+1

**Comment by drewhemm** [comment]
Would it make any difference if the pods concerned had a livenessProbe or readinessProbe?

It is possible to override the restartPolicy for GH ARC runner pods by using the [template](https://github.com/actions/actions-runner-controller/blob/master/charts/gha-runner-scale-set/values.yaml#L274) Helm chart value, but I can't imagine that would be a good idea. One of the main benefits of the GH ARC is that runner pods are single-use only.

**Comment by jkleinlercher** [comment]
> Maybe we can implement an annotation to stop this behaviour? Then we wouldn't have to break anything current?

I would vote for this solution

**Comment by RoelofKuijpers** [comment]
I have created a PR containing an implementation of the above solution direction. [PR 709](https://github.com/argoproj/gitops-engine/pull/709).

**Comment by Liammarwood** [comment]
Any progress on merging this?

**Comment by nitishfy** [comment]
TLDR:

The Argo CD application shows the App health as _Progressing_ when it deploys a Pod that has the `restartPolicy` set to `Never` or `OnFailure` (see [Kubernetes docs for restart policy](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)). Generally, these resources that have `restartPolicy` set to `Never` behave like a job. This is [intended behavior](https://github.com/argoproj/argo-cd/issues/7933#issuecomment-993949395), since Jobs are commonly used for sync hooks and must finish before an Application is considered Healthy. A workaround that is suggested here is to have an annotation something like `argocd.argoproj.io/ignore-restart-policy` which, when set on the resource, will ignore the restart policy and return the health status provided the pod is running. 

Here's the [PR](https://github.com/argoproj/gitops-engine/pull/709) that implements that.

**Comment by agaudreault** [comment]
If you do not want a specific resource to affect the overall Application health, then the resource should have a `argocd.argoproj.io/ignore-healthcheck: 'true'` annotation. The resource itself will still be progressing. Otherwise, the restartPolicy should be to Always for long running Pod.

**Comment by gcezaralmeida** [comment]
Hey Guys, any workarounds, my app is showing as Healthy, however, my virt-launcher pods are stuck in progress. Everything seems fine, but this weird status for the pods that is not nice.

**Comment by unlightable** [comment]
Another example of this occurring is an Agones GameServer's with `restartPolicy: never`.
Would be nice to have an annotation fixing it!

**Comment by fuzolan** [comment]
Annotation would be really nice

**Comment by utdrmac** [comment]
Hey all. Just started looking at ArgoCD a week ago and this 'progressing' issue really frustrated me. We have many pods that are technically "long running" but we don't set the restart policy, as we would rather the pod stop if there's a failure with the app running inside.
We _thought_ that by adding a proper readiness check to our pods, argo would see this, but that is not the case. I do see now in the comments how restart:never is ignored by argo. If argo cannot simply trust the readiness check, then having this annotation exception would be very nice for us.

**Comment by RoelofKuijpers** [comment]
I created this PR to address the issue above: https://github.com/argoproj/gitops-engine/pull/709. I still need to review the comments @agaudreault left. I'll follow up shortly.

**Comment by jskrill** [comment]
Since it's been a while that there was any progress on this issue and now gitops-engine is rolled into the argo-cd repo, I've redone this PR. I've incorporated feedback from the comments (especially regarding the interaction with hooks), and you can see the new PR here at https://github.com/argoproj/argo-cd/pull/28249.

With respect to whether ```argocd.argoproj.io/ignore-healthcheck: 'true'``` solves this issue (commented on the other PR), I'd say it does not. There are two reasons for this. The first is that `argocd.argoproj.io/ignore-healthcheck` doesn't actually fix the pod health status which will be stuck in Progressing, it just keeps the Progressing pod from effecting the roll-up application status. End-users using operators that own these types of pods will always see the pod as Progressing which is misleading from a UI perspective when the pod is actually in a steady state. The other reason `argocd.argoproj.io/ignore-healthcheck` isn't sufficient is that we don't actually want to ignore the health check. If one of these pods is legitimately Progressing/Failing/etc, that should still be reflected. What we really want is to adjust the health check logic without disrupting the existing ArgoCD hooks logic and that's what this PR accomplishes via an opt-in annotation.
